### DISCUSSION TITLE
代理使用 direct 直连 cpu 占用很高

### DISCUSSION THREAD

**Discussion by lemos1235** [xjasonlyu/tun2socks#143] [Q&A]
我是在 windows 下使用，关于路由循环问题，参考 https://github.com/iamwwcposts/articles/issues/57

新增加的代码如下：
![image](https://user-images.githubusercontent.com/23135552/176091169-e6feea27-0a01-4b62-8fe6-d561b0b4d025.png)

在使用-proxy 指定代理的时候，如果协议设置为 socks:// 是正常的，但是改成 direct:// 发现就不太正常了，cpu 占用很高

控制台会一致打印这样的数据

![image](https://user-images.githubusercontent.com/23135552/176090892-63019362-d901-41ff-9967-2855996fb9ec.png)

麻烦作者解答一下

**Comment by xjasonlyu** [comment]
这个其实我也不是很清楚怎么在windows下解决policy routing问题。

**Reply by lemos1235** [reply]
读了 clash 源码，发现了[解决方法 ](https://github.com/Dreamacro/clash/blob/c231fd14666d6ea05d6a75eaba6db69f9eee5ae9/component/dialer/bind_others.go#L78)

![image](https://user-images.githubusercontent.com/23135552/176343577-6910719f-2240-47ab-93bf-062ed63d28ff.png)

即将 address 改成 local address（这里我当前的局域网IP是 192.168.0.104）就完全没问题了。

**Reply by lemos1235** [reply]
还有一点，如果代理服务器在本地，比如代理地址为 socks5://127.0.0.1:10808，则需要手动将该程序对应的流量排除，让其走 direct。否则还是会产生循环。

**Reply by larry19840909** [reply]
[lemos1235](https://github.com/lemos1235) 我想问下，你这样调整以后，有没有在direct.go里将netbios类型的请求过滤掉，wintun创建成功后，会有大量的netbios的请求，如果不把169.xxx.xxx.xxx或者目标端口137,138的过滤掉依然会产生回环。

**Reply by rurirei** [reply]
> [lemos1235](https://github.com/lemos1235) 我想问下，你这样调整以后，有没有在direct.go里将netbios类型的请求过滤掉，wintun创建成功后，会有大量的netbios的请求，如果不把169.xxx.xxx.xxx或者目标端口137,138的过滤掉依然会产生回环。

how to disable the netbios?

**Reply by rurirei** [reply]
> 读了 clash 源码，发现了[解决方法 ](https://github.com/Dreamacro/clash/blob/c231fd14666d6ea05d6a75eaba6db69f9eee5ae9/component/dialer/bind_others.go#L78)
> 
> ![image](https://user-images.githubusercontent.com/23135552/176343577-6910719f-2240-47ab-93bf-062ed63d28ff.png)
> 
> 即将 address 改成 local address（这里我当前的局域网IP是 192.168.0.104）就完全没问题了。

you place the lc.Control function in either inbound or outbound listener for udp conn?

**Comment by rurirei** [comment]
met this help wanted.
```
dial tcp: The requested address is not valid in its context.
```
```golang
dialer := &net.Dialer{
	Control: func(network, address string, c syscall.RawConn) error {
             return c.Control(func(fd uintptr) {
	             bindSocketToInterface6(windows.Handle(fd), uint32(ifce.Index))
	     })
        },
}
```
